src/main/webapp/app/entities/result/result.component.ts

```diff
@@ -51,54 +71,108 @@ export class ResultComponent implements OnInit, OnChanges {
-
-            if (this.participation.results && this.participation.results.length > 0) {
-                if (exercise && exercise.type === ExerciseType.MODELING) {
-                    // sort results by completionDate descending to ensure the newest result is shown
-                    // this is important for modeling exercises since students can have multiple tries
-                    // think about if this should be used for all types of exercises
-                    this.participation.results.sort((r1: Result, r2: Result) => {
-                        if (r1.completionDate! > r2.completionDate!) {
-                            return -1;
-                        }
-                        if (r1.completionDate! < r2.completionDate!) {
-                            return 1;
-                        }
-                        return 0;
-                    });
-                }
+            if (this.hasParticipationResults()) {
+                const result = this.getLatestResult(this.participation.results);
+
                 // Make sure result and participation are connected
-                this.result = this.participation.results[0];
+                this.result = result;
                 this.result.participation = this.participation;
             }
 
-            this.init();
+            this.evaluate();
         }
     }
 
-    init() {
-        if (this.result && (this.result.score || this.result.score === 0) && (this.result.rated === true || this.result.rated == null || this.showUngradedResults)) {
+    evaluate() {
+        this.evaluateTemplateStatus();
+
+        if (this.result && (this.result.score || this.result.score === 0) && (this.result.rated || this.result.rated == null || this.showUngradedResults)) {
             this.textColorClass = this.getTextColorClass();
             this.hasFeedback = this.getHasFeedback();
             this.resultIconClass = this.getResultIconClass();
             this.resultString = this.buildResultString();
             this.resultTooltip = this.buildResultTooltip();
+        } else if (this.templateStatus === ResultTemplateStatus.LATE) {
+            this.textColorClass = 'result-gray';
+            this.resultIconClass = this.getResultIconClass();
         } else {
             // make sure that we do not display results that are 'rated=false' or that do not have a score
             this.result = null;
         }
     }
 
     ngOnChanges(changes: SimpleChanges) {
-        if (changes.participation || changes.result) {
-            this.ngOnInit();
+        if (changes.participation || changes.result || changes.participation) {
+            this.init();
+        }
+    }
+
+    evaluateTemplateStatus() {
+        if (this.isModelingOrTextOrFileUpload() && this.participation && this.participation.exercise) {
+            const assessmentDueDate = this.dateAsMoment(this.participation.exercise!.assessmentDueDate!);
+            if (this.isSubmissionInDueTime()) {
```

Comment:
Try not to nest the too many ifs, normally you should try to avoid nesting at all.
What does the control flow at this stage mean? So when would we even enter this branch of the code?

Reply:
I've moved the evaluation for text/modeling/file-upload exercises and for qiuz/programming exercises to separate methods and I've also flattened out the nested if statements. In addition, I have documented each branch.